model/infinispan/src/main/java/org/keycloak/infinispan/module/certificates/CertificateReloadManager.java

```diff
@@ -0,0 +1,283 @@
+    @Start
+    public void start() {
+        logger.debug("Starting JGroups certificate reload manager");
+        notifier.addListener(this);
+        reloadCertificate();
+        certificateHolder.setExceptionHandler(this::onInvalidCertificate);
+    }
+
+    @Override
+    @Stop
+    public void stop() {
+        logger.debug("Stopping JGroups certificate reload manager");
+        notifier.removeListener(this);
+        lock.lock();
+        try (lock) {
+            if (scheduledFuture == null) {
+                return;
+            }
+            scheduledFuture.cancel(true);
+        }
+    }
+
+    /**
+     * Creates and reload a new certificate.
+     */
+    public void rotateCertificate() {
+        logger.debug("Rotate JGroups certificate");
+        lock.lock();
+        try (lock) {
+            KeycloakModelUtils.runJobInTransaction(sessionFactory, this::replaceCertificateInTransaction);
+            sendReloadNotification();
+        } catch (RuntimeException e) {
+            logger.warn("Failed to rotate JGroups certificate", e);
+            retry(this::rotateCertificate, "retry-rotate");
+        }
+    }
+
+    /**
+     * Reloads the certificate from storage.
+     */
+    public void reloadCertificate() {
+        logger.debug("Reload JGroups Certificate");
+        lock.lock();
+        try (lock) {
+            var maybeCrt = KeycloakModelUtils.runJobInTransactionWithResult(sessionFactory, CertificateReloadManager::loadCertificateInTransaction);
+            if (maybeCrt.isEmpty()) {
+                return;
+            }
+            var crt = JGroupsCertificate.fromJson(maybeCrt.get());
+            certificateHolder.useCertificate(crt);
+        } catch (GeneralSecurityException | IOException e) {
+            logger.warn("Failed to reload JGroups certificate", e);
+            retry(this::reloadCertificate, "retry-reload");
+        } finally {
+            scheduleNextRotation();
+        }
+    }
+
+    @ViewChanged
+    @Merged
+    public void onViewChanged(ViewChangedEvent event) {
+        logger.debug("On view changed");
+        // probably a waste to reload, but if we have a partition, we reload the most recent certificate stored.
+        reloadCertificate();
+    }
+
+    // testing purpose
+    public JGroupsCertificate currentCertificate() {
+        return certificateHolder.getCertificateInUse();
+    }
+
+    // testing purpose
+    public void setRotationSeconds(long seconds) {
+        this.rotationSeconds = seconds;
+    }
+
+    // testing purpose
+    public boolean isCoordinator() {
+        return cacheManager.isCoordinator();
+    }
+
+    // testing purpose
+    public boolean hasRotationTask() {
+        lock.lock();
+        try (lock) {
+            return scheduledFuture != null;
+        }
+    }
+
+    private void onInvalidCertificate() {
+        logger.debug("On certificate exception");
+        blockingManager.runBlocking(this::reloadCertificate, "invalid-certificate");
```

Comment:
I'm wondering if this is a vector for a DDoS attack. Trying to establish a connection will trigger a database read. Add some rate limit, as 1 request per minute or so?

Reply:
I think this is still cheap: There will be a DB read, but the contents will have been cached by the database. 
If the code reads the same alias before, the trust manager is left untouched. The JGroups port is also an internal one only, not reachable by the public by design. 

So would merge it as is. 

As a follow-up, we could do a rate limiting to have done like once per second, or ensure that at max one additional execution is scheduled while one is already running. Any excess requests could then be safely discarded. Again, this would be IMHO a follow-up.

Reply:
I'm ok either way. I want to add a test with partition, too, at some point. I can do it in a follow-up if we wish to review/merge this today.